File: app/src/main/java/com/amazon/tv/leanbacklauncher/notifications/RecommendationCardView.java
package com.amazon.tv.leanbacklauncher.notifications;

import android.content.Context;
import android.content.pm.PackageManager.NameNotFoundException;
import android.content.res.Resources.NotFoundException;
import android.graphics.Bitmap;

import androidx.annotation.NonNull;

import com.amazon.tv.leanbacklauncher.PackageResourceCache;
import com.amazon.tv.tvrecommendations.TvRecommendation;
import com.bumptech.glide.request.target.SizeReadyCallback;

public class RecommendationCardView extends RecommendationView {
    private TvRecommendation mRecommendation;
    private PackageResourceCache mResourceCache;

    public RecommendationCardView(Context context) {
        this(context, PackageResourceCache.getInstance(context));
    }

    RecommendationCardView(Context context, PackageResourceCache resourceCache) {
        super(context);
        this.mResourceCache = resourceCache;
    }

    public void bindInfoAreaTitleAndContent() {
        this.mTitleView.setText(this.mRecommendation.getTitle());
        this.mContentView.setText(this.mRecommendation.getText());
    }

    public Bitmap getContentImage() {
        return this.mRecommendation != null ? this.mRecommendation.getContentImage() : null;
    }

    public int getDataWidth() {
        return this.mRecommendation != null ? this.mRecommendation.getWidth() : 0;
    }

    public int getDataHeight() {
        return this.mRecommendation != null ? this.mRecommendation.getHeight() : 0;
    }

    public boolean hasProgress() {
        return this.mRecommendation != null && this.mRecommendation.hasProgress();
    }

    public int getProgress() {
        return this.mRecommendation != null ? this.mRecommendation.getProgress() : 0;
    }

    public void setRecommendation(TvRecommendation rec, boolean updateImage) {
        this.mRecommendation = rec;
        int color = rec.getColor();
        if (color == 0) {
            color = this.mInfoAreaDefaultColor;
        }
        this.mInfoAreaColor = color;


        bindProgressBar(this.mRecommendation.getProgressMax(), this.mRecommendation.getProgress());
        bindSourceName(this.mRecommendation.getSourceName(), this.mRecommendation.getPackageName());
        bindBadge();
        bindContentDescription(this.mRecommendation.getTitle(), this.mRecommendation.getSourceName(), this.mRecommendation.getText());
        onBind();
    }


    public String getWallpaperUri() {
        return this.mRecommendation.getBackgroundImageUri();
    }

    public String getSignature() {
        if (this.mSignature == null) {
            this.mSignature = "" + this.mRecommendation.getTitle() + this.mRecommendation.getText() + this.mRecommendation.getSourceName();
        }
        return this.mSignature;
    }

    protected void bindBadge() {
        try {
            // todo
            bindBadge(this.mResourceCache.getDrawable(this.mRecommendation.getPackageName(), this.mRecommendation.getBadgeIcon()));
            return;
        } catch (NameNotFoundException e) {
        } catch (NotFoundException e2) {
        }
        bindBadge(null);
    }

    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        boolean isLayoutRtl = true;
        super.onLayout(changed, left, top, right, bottom);
        int width = right - left;
        int height = bottom - top;
        if (getLayoutDirection() != 1) {
            isLayoutRtl = false;
        }
        layoutMainImage(width);
        layoutProgressBar(width);
        layoutSourceName(width, isLayoutRtl);
        layoutBadgeIcon(width, height, isLayoutRtl);
        layoutExpandedInfoArea(width, isLayoutRtl);
    }

    @Override
    public void removeCallback(@NonNull SizeReadyCallback cb) {

    }
}

File: app/src/main/java/com/amazon/tv/leanbacklauncher/notifications/RecommendationCardView.kt
package com.amazon.tv.leanbacklauncher.notifications

import android.content.Context
import android.graphics.Bitmap
import com.amazon.tv.leanbacklauncher.PackageResourceCache
import com.amazon.tv.tvrecommendations.TvRecommendation
import com.bumptech.glide.request.target.SizeReadyCallback

class RecommendationCardView internal constructor(context: Context?, private val mResourceCache: PackageResourceCache?) : RecommendationView(context!!) {
    private var mRecommendation: TvRecommendation? = null

    constructor(context: Context?) : this(context, PackageResourceCache.getInstance(context))

    public override fun bindInfoAreaTitleAndContent() {
        mTitleView!!.text = mRecommendation!!.title
        mContentView!!.text = mRecommendation!!.text
    }

    override val contentImage: Bitmap?
        get() = if (mRecommendation != null) mRecommendation!!.contentImage else null
    override val dataWidth: Int
        get() = if (mRecommendation != null) mRecommendation!!.width else 0
    override val dataHeight: Int
        get() = if (mRecommendation != null) mRecommendation!!.height else 0

    public override fun hasProgress(): Boolean {
        return mRecommendation != null && mRecommendation!!.hasProgress()
    }

    val progress: Int
        get() = if (mRecommendation != null) mRecommendation!!.progress else 0

    fun setRecommendation(rec: TvRecommendation, updateImage: Boolean) {
        mRecommendation = rec
        var color = rec.color
        if (color == 0) {
            color = mInfoAreaDefaultColor
        }
        mInfoAreaColor = color
        bindProgressBar(mRecommendation!!.progressMax, mRecommendation!!.progress)
        bindSourceName(mRecommendation!!.sourceName, mRecommendation!!.packageName)
        bindBadge()
        bindContentDescription(mRecommendation!!.title, mRecommendation!!.sourceName, mRecommendation!!.text)
        onBind()
    }

    override val wallpaperUri: String?
        get() = mRecommendation!!.backgroundImageUri

    override val signature: String?
        get() {
            if (mSignature == null) {
                mSignature = "" + mRecommendation!!.title + mRecommendation!!.text + mRecommendation!!.sourceName
            }
            return mSignature
        }

    override fun bindBadge() {
        try {
            // TODO
            bindBadge(mResourceCache!!.getDrawable(mRecommendation!!.packageName, mRecommendation!!.badgeIcon))
            return
        } catch (e: Exception) {
            e.printStackTrace()
        }
        bindBadge(null)
    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        var isLayoutRtl = true
        super.onLayout(changed, left, top, right, bottom)
        val width = right - left
        val height = bottom - top
        if (layoutDirection != 1) {
            isLayoutRtl = false
        }
        layoutMainImage(width)
        layoutProgressBar(width)
        layoutSourceName(width, isLayoutRtl)
        layoutBadgeIcon(width, height, isLayoutRtl)
        layoutExpandedInfoArea(width, isLayoutRtl)
    }

    override fun removeCallback(cb: SizeReadyCallback) {}
}